text/0819-public-router-service.md

```diff
@@ -0,0 +1,242 @@
+---
+Stage: Accepted
+Start Date: 2022-05-12
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/819
+---
+
+<!--- 
+Directions for above: 
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Public `RouterService` import
+
+## Summary
+
+This RFC proposes adding a public import for [`RouterService`](https://api.emberjs.com/ember/release/classes/RouterService)
+making it extensible by addons, removing the external link assertion from `<LinkTo />` and rebuilding the `<LinkTo />` component
+to rely entirely on the public routing service instead of the internal `-router`.
+
+## Motivation
+
+Today, users of [`ember-engines`](https://github.com/ember-engines/ember-engines) are required to utilize separate
+custom components, helpers and services for linking. This leaves the experience feeling sub-par, in addition to introducing
+extra maintenance friction attempting to keep these primitives aligned with those in `ember-source`.
+
+By providing a public import for `RouterService`, `ember-engines` would be able to provide its own router service
+implementation extending from the base implementation. Combined with minor changes to the engines-specific behavior of
+`link-to`, this change would allow us to fill the existing DX gap when it comes to navigating between an Engine
+and a host App *and* allow engines to eliminate use of the legacy components required to support 4.x versions of Ember.
+
+These proposed changes would further allow `ember-engines` to deprecate the `<LinkToExternal />`
+component provided by `ember-engines` package. As a result, there will be less
+engine specific concepts and code would be more portable between regular applications and engines.
+
+Once this RFC gets implemented, following will be possible in user defined Ember Engines:
+
+* links to external routes in templates within the engine rendered via `<LinkTo />` component:
+
+  ```hbs
+  <LinkTo @route="external-route">Go Somewhere</LinkTo>
+  ```
+
+* using `RouterService` methods for redirects to the external routes:
+
+  ```javascript
+  # addon/components/my-component.js
+  export default class extends Component {
+    @service router;
+    @action clickLink () {
+      this.router.transitionTo('external-route');
```

Comment:
Also, I'd like to have something like that to don't lose the explicit outside context on engines 

```ts
import { external } from 'ember-engines/helper'
...
@action 
clickLink () {
   this.router.transitionTo(external('external-route'));
 }
```

Reply:
It feels unnecessary to impose an extra function call on users to mark something as external when it has already been defined as external in their router configuration (see also comments above on `external.`)

Reply:
It depends, I kinda agree with you that mapping those things on the router configuration clarifies things on a certain level, but I love readable and explicit code. How would you realize just looking for a ton of `transitionTo` in your code? imagine you have at least 50 occurrences how many are linking outside context? Correct me if I'm wrong, IMO is necessary back in forth on the router configuration to realize those things.